internal/kafka/internal/handlers/validation.go

```diff
@@ -51,45 +85,113 @@ func ValidateKafkaClusterNameIsUnique(name *string, kafkaService services.KafkaS
 		// Set Cloud Provider default if not received in the request
 		supportedProviders := providerConfig.ProvidersConfig.SupportedProviders
-		if kafkaRequest.CloudProvider == "" {
-			defaultProvider, _ := supportedProviders.GetDefault()
-			kafkaRequest.CloudProvider = defaultProvider.Name
-		}
 
+		defaultProvider, _ := supportedProviders.GetDefault()
+		providerName := arrays.FirstNonEmptyOrDefault(defaultProvider.Name, kafkaRequest.CloudProvider)
 		// Validation for Cloud Provider
-		provider, providerSupported := supportedProviders.GetByName(kafkaRequest.CloudProvider)
+		provider, providerSupported := supportedProviders.GetByName(providerName)
 		if !providerSupported {
 			return errors.ProviderNotSupported("provider %s is not supported, supported providers are: %s", kafkaRequest.CloudProvider, supportedProviders)
 		}
 
-		// Set Cloud Region default if not received in the request
-		if kafkaRequest.Region == "" {
-			defaultRegion, _ := provider.GetDefaultRegion()
-			kafkaRequest.Region = defaultRegion.Name
+		// Validation for Cloud Region
+		if kafkaRequest.Region != "" { // if region is empty, default region will be chosen, so no validation is needed
+			regionSupported := provider.IsRegionSupported(kafkaRequest.Region)
+			if !regionSupported {
+				return errors.RegionNotSupported("region %s is not supported for %s, supported regions are: %s", kafkaRequest.Region, kafkaRequest.CloudProvider, provider.Regions)
+			}
 		}
 
-		// Validation for Cloud Region
-		regionSupported := provider.IsRegionSupported(kafkaRequest.Region)
-		if !regionSupported {
-			return errors.RegionNotSupported("region %s is not supported for %s, supported regions are: %s", kafkaRequest.Region, kafkaRequest.CloudProvider, provider.Regions)
+		claims, ok := validatorContext[CLAIMS]
+		if !ok {
+			err := ValidateKafkaClaims(validatorContext, ctx)()
+			if err != nil {
+				return err
+			}
+			claims = validatorContext[CLAIMS]
 		}
 
+		owner := auth.GetUsernameFromClaims(claims.(jwt.MapClaims))
+		organisationId := auth.GetOrgIdFromClaims(claims.(jwt.MapClaims))
+
 		// Validate Region/InstanceType
-		instanceType, err := (*kafkaService).AssignInstanceType(kafkaRequest)
+		instanceType, err := (*kafkaService).AssignInstanceType(owner, organisationId)
 		if err != nil {
 			return errors.NewWithCause(errors.ErrorGeneral, err, "error assigning instance type: %s", err.Error())
 		}
 
-		region, _ := provider.Regions.GetByName(kafkaRequest.Region)
+		var region config.Region
+		if kafkaRequest.Region == "" {
+			region, _ = provider.GetDefaultRegion()
+		} else {
+			region, _ = provider.Regions.GetByName(kafkaRequest.Region)
+		}
+
 		if !region.IsInstanceTypeSupported(config.InstanceType(instanceType)) {
 			return errors.InstanceTypeNotSupported("instance type '%s' not supported for region '%s'", instanceType.String(), region.Name)
 		}
+
+		validatorContext[CLOUD_PROVIDER] = provider.Name
+		validatorContext[REGION] = region.Name
```

Comment:
This is side effect..

Reply:
There are a lot of these in the validators..

Reply:
Sure it is. But, IMO, it makes the code simpler, while the proposed approach makes it unreadable.
Anyway, looks like I'm alone in this battle, so I will follow your suggestions :)

Reply:
> Sure it is

it is not different from what we've at the moment which is what the JIRA is proposing to get rid of.

> But, IMO, it makes the code simpler

It is hard to reason about the code and violates the principle of least surprises ;-) i.e I'd not expect the validator to have side effects and its the last place someone will look to see where the initialization of a `foo` or `bar` field is being made.

> so I will follow your suggestions

Thanks.

Reply:
> it is not different from what we've at the moment which is what the JIRA is proposing to get rid of.
It is very different. Currently the validator expects some data to be ready and the order of validators is important.
With this change the validator does not expect anything and doesn't care about the order. You can even remove validators and it will work anyway.

> someone will look to see where the initialization of a foo or bar field is being made.

Nope . As I stated before, validator doesn't have to rely to the content of the context. They can check if a value is there, but if it is not there, they have to produce that themselves.